### ISSUE TITLE
New YAML engine writes invalid yaml arrays

### ISSUE THREAD

**Issue by nesl247** [pulumi/pulumi#6679]
With the new yaml engine in 2.24.0, previously written valid yaml is rewritten in invalid yaml when running `pulumi config set`. The invalid yaml are arrays, however, this is triggered by updating anything in the file. It's only necessary to trigger pulumi to rewrite the file to cause the problem.

## Expected behavior
This is a subsection of our `Pulumi.staging.yaml` file.

```yaml
    unvailablePods:
      configuration:
        notifyTargets:
          - '@some-target'
        priorities:
          critical: P3
```

## Current behavior
This is a subsection of our `Pulumi.staging.yaml` file.

```yaml
    unvailablePods:
      configuration:
        notifyTargets:
        - @some-target
        priorities:
          critical: P3
```

## Steps to reproduce

1. Use pulumi 2.24.0
2. Have a config that has an array of strings
3. Use `pulumi config set` to change any value in the configuration (it does not need to be related to the array at all)
4. The array is rewritten in invalid yaml

## Context (Environment)
This occurs using pulumi 2.24.0.

## Affected feature

Our deployment pipeline got broken by this as pulumi is writing invalid yaml. We had to revert to a fixed version pointing to  the previous release.

**Comment by mikhailshilkov** [comment]
@lblackstone this sounds like a nasty issue, could you please take a look?

**Comment by lblackstone** [comment]
This must be a bug in the new YAML package because https://github.com/pulumi/pulumi/pull/6489 hasn't merged yet. I'll see about getting this fixed today.

**Comment by lblackstone** [comment]
We decided to revert the package change for now until we can make sure this is fixed. We'll cut a patch release ASAP.

**Comment by stack72** [comment]
Hi @nesl247 

As a follow up to what @lblackstone said, I am cutting v2.24.1 release right now

Paul

**Comment by stack72** [comment]
Hey @nesl247 this is closed via #6689

**Comment by nesl247** [comment]
Thanks @stack72 and @lblackstone.
